src/native/cipher.rs

```diff
@@ -196,12 +228,23 @@ fn _decrypt_in<P: PoolParams>(eta:Num<P::Fr>, mut memo:&[u8], params:&P)->Option
     let note = (0..nozero_notes_num).map(|i| {
         let a_pub = EdwardsPoint::subgroup_decompress(Num::deserialize(&mut memo).ok()?, params.jubjub())?;
         let ecdh = a_pub.mul(eta.to_other_reduced(), params.jubjub());
-        let key = keccak256(&ecdh.x.try_to_vec().unwrap());
+        
+        let key = {
+            let mut x: [u8; 32] = [0; 32];
+            ecdh.x.serialize(&mut &mut x[..]).unwrap();
+            keccak256(&x)
+        };
 
         let ciphertext = buf_take(&mut memo, note_size+constants::POLY_1305_TAG_SIZE)?;
-        let text = symcipher_decode(&key, ciphertext)?;
-        let note = Note::try_from_slice(&text).ok()?;
-        if note.hash(params) != note_hash[i] {
+        let plain = symcipher_decode::<NOTE_HEAPLESS_SIZE>(&key, ciphertext)?;
+        let note = Note::try_from_slice(plain.as_slice()).ok()?;
+
+        let note_hash = {
+            let note_hash = &mut &note_hashes[i * num_size..(i + 1) * num_size];
+            Num::deserialize(note_hash).ok()?
```

Comment:
We can call deserialize directly from `note_hashes` here (`Num::deserialize(&mut note_hashes).ok()?`) if we mark it as mutable to prevent index calculation for array slices. Not really necessary, but simplifies the logic a bit.

Reply:
I think we actually can't do so here. In this loop, we can skip some foreign notes so `note_hashes` will not be shifted correctly and deserialized value in this case will be wrong. We should be able to do it in `decrypt out` but the current solution seems a bit more reliable to me in cases when there are some problems in memo. What do you think?

Reply:
Agree. When a decryption fails we will continue the loop skipping deserialization. So, it won't work correctly.